File: servers/go/httpapi/passkey_issuer_test.go
package httpapi

import (
	"bytes"
	"context"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"errors"
	"net/http/httptest"
	"path/filepath"
	"testing"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/go-webauthn/webauthn/protocol/webauthncbor"
	"github.com/mattydsmith/passkey/servers/go/auth"
	"github.com/mattydsmith/passkey/servers/go/passkey"
	"github.com/mattydsmith/passkey/servers/go/storage"
)

func TestPasskeyIssuerVerifiedHTTP(t *testing.T) {
	for _, mode := range []string{"success", "denied", "unavailable", "failure", "empty", "wrong-user", "invalid-signature", "default"} {
		t.Run(mode, func(t *testing.T) {
			s, err := storage.OpenSQLite(filepath.Join(t.TempDir(), "auth.db"))
			if err != nil {
				t.Fatal(err)
			}
			t.Cleanup(func() {
				if err := s.Close(); err != nil {
					t.Error(err)
				}
			})
			key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
			if err != nil {
				t.Fatal(err)
			}
			public, err := webauthncbor.Marshal(map[int]any{1: 2, 3: -7, -1: 1, -2: key.X.FillBytes(make([]byte, 32)), -3: key.Y.FillBytes(make([]byte, 32))})
			if err != nil {
				t.Fatal(err)
			}
			id := []byte("synthetic-passkey")
			at := time.Now()
			if err := s.CreatePasskey(storage.Passkey{CredentialID: id, PublicKey: public, UserID: "synthetic-user", CreatedAt: at}); err != nil {
				t.Fatal(err)
			}
			calls := 0
			cfg := Config{Storage: s, RPID: "example.com", Origins: []string{"https://example.com"}, SessionCookieName: "session", Now: func() time.Time { return at }}
			if mode != "default" {
				cfg.PasskeySignIn = func(ctx context.Context, in passkey.SignInInput) (passkey.SignInResult, error) {
					calls++
					if ctx != in.Request.Context() || in.UserID != "synthetic-user" || !bytes.Equal(in.CredentialID, id) || !bytes.Equal(in.PublicKey, public) || in.SignCount != 1 || in.SessionTTL != 30*24*time.Hour || !in.Now().Equal(at) || in.Request.RemoteAddr != "192.0.2.11:1234" || in.UserAgent == nil || *in.UserAgent != "passkey-test" || in.IP == nil || *in.IP != "spoofed" {
						t.Fatal("host did not receive verified proof/policy/request")
					}
					switch mode {
					case "unavailable":
						return passkey.SignInResult{}, auth.ErrSessionUnavailable
					case "denied":
						return passkey.SignInResult{}, passkey.ErrSignInDenied
					case "failure":
						return passkey.SignInResult{}, errors.New("private host failure")
					case "empty":
						return passkey.SignInResult{}, nil
					case "wrong-user":
						return passkey.SignInResult{SessionToken: "host-token", UserID: "different-user"}, nil
					}
					return passkey.SignInResult{SessionToken: "host-token", UserID: in.UserID}, nil
				}
			}
			router := chi.NewRouter()
			if err := Mount(router, cfg); err != nil {
				t.Fatal(err)
			}
			start := httptest.NewRecorder()
			router.ServeHTTP(start, httptest.NewRequest("POST", "/auth/passkey/sign-in/start", nil))
			var challenge struct {
				SignInID string `json:"signInId"`
				Options  struct {
					Challenge string `json:"challenge"`
				} `json:"options"`
			}
			if err := json.Unmarshal(start.Body.Bytes(), &challenge); err != nil || challenge.Options.Challenge == "" {
				t.Fatalf("start failed status=%d", start.Code)
			}
			clientData, _ := json.Marshal(map[string]any{"type": "webauthn.get", "challenge": challenge.Options.Challenge, "origin": "https://example.com"})
			rpHash := sha256.Sum256([]byte("example.com"))
			authData := append(rpHash[:], byte(5))
			authData = append(authData, make([]byte, 4)...)
			binary.BigEndian.PutUint32(authData[33:], 1)
			clientHash := sha256.Sum256(clientData)
			signed := append(append([]byte{}, authData...), clientHash[:]...)
			digest := sha256.Sum256(signed)
			signature, err := ecdsa.SignASN1(rand.Reader, key, digest[:])
			if err != nil {
				t.Fatal(err)
			}
			if mode == "invalid-signature" {
				signature[len(signature)-1] ^= 1
			}
			b64 := base64.RawURLEncoding.EncodeToString
			body, _ := json.Marshal(map[string]any{"signInId": challenge.SignInID, "credential": map[string]any{"id": b64(id), "rawId": b64(id), "type": "public-key", "response": map[string]any{"clientDataJSON": b64(clientData), "authenticatorData": b64(authData), "signature": b64(signature), "userHandle": b64([]byte("synthetic-user"))}, "clientExtensionResults": map[string]any{}}})
			req := httptest.NewRequest("POST", "/auth/passkey/sign-in/finish", bytes.NewReader(body))
			req.RemoteAddr = "192.0.2.11:1234"
			req.Header.Set("User-Agent", "passkey-test")
			req.Header.Set("X-Forwarded-For", "spoofed")
			rec := httptest.NewRecorder()
			router.ServeHTTP(rec, req)
			wantStatus, wantCalls := 500, 1
			switch mode {
			case "success":
				wantStatus = 200
			case "default":
				wantStatus = 200
				wantCalls = 0
			case "unavailable":
				wantStatus = 503
			case "denied":
				wantStatus = 401
			case "invalid-signature":
				wantStatus = 401
				wantCalls = 0
			}
			if rec.Code != wantStatus || calls != wantCalls {
				t.Fatalf("mode=%s status=%d calls=%d", mode, rec.Code, calls)
			}
			stored, err := s.GetPasskey(id)
			if err != nil {
				t.Fatal(err)
			}
			sessions, err := s.ListSessions("synthetic-user")
			if err != nil {
				t.Fatal(err)
			}
			if mode == "default" {
				if stored.SignCount != 1 || len(sessions) != 1 {
					t.Fatal("legacy issuer did not persist")
				}
			} else if stored.SignCount != 0 || len(sessions) != 0 {
				t.Fatal("host path invoked default persistence")
			}
			if mode == "success" || mode == "default" {
				if len(rec.Result().Cookies()) != 2 {
					t.Fatal("success missing cookies")
				}
			} else if len(rec.Result().Cookies()) != 0 || bytes.Contains(rec.Body.Bytes(), []byte("host-token")) || bytes.Contains(rec.Body.Bytes(), []byte("private host failure")) {
				t.Fatal("failure emitted credentials or private errors")
			}
			again := httptest.NewRecorder()
			router.ServeHTTP(again, httptest.NewRequest("POST", "/auth/passkey/sign-in/finish", bytes.NewReader(body)))
			if again.Code == 200 || calls != wantCalls {
				t.Fatal("spent challenge replayed issuer")
			}
		})
	}
}

File: packages/core/tests/passkey-issuer.test.ts
// Plumbing proof with the WebAuthn verifier explicitly stubbed. The Go HTTP
// counterpart uses signed assertions against the real verifier.
import { describe, it, expect, vi, beforeEach } from "vitest";
import { createAuth } from "../src/auth.js";
import { AuthError } from "../src/errors.js";
import type { PasskeySignIn } from "../src/types.js";
import { createHarness } from "./setup.js";
import { insertPasskey, getPasskeyByCredentialId } from "../src/storage/passkeys.js";
vi.mock("@simplewebauthn/server", async (importOriginal) => ({
  ...await importOriginal<typeof import("@simplewebauthn/server")>(),
  verifyAuthenticationResponse: vi.fn(),
}));
import { verifyAuthenticationResponse } from "@simplewebauthn/server";

const credentialId = new Uint8Array([1,2,3]);
const publicKey = new Uint8Array([4,5,6]);
const credential = {
 id:"AQID", rawId:"AQID", type:"public-key" as const,
 response:{clientDataJSON:"e30", authenticatorData:"e30", signature:"e30", userHandle:null},
 clientExtensionResults:{},
};

describe("host passkey issuer",()=>{
 beforeEach(()=>vi.resetAllMocks());
 for(const mode of ["success","direct","denied","unavailable","failure","empty","wrong-user","invalid","default"]){
  it(mode,async()=>{
   const h=createHarness();
   try {
    insertPasskey(h.db,{credentialId,publicKey,userId:"user",signCount:0,transports:null,aaguid:null,deviceName:null,createdAt:h.clock.now,lastUsedAt:null});
    vi.mocked(verifyAuthenticationResponse).mockResolvedValue({verified:mode!=="invalid",authenticationInfo:{newCounter:1}} as any);
    const request=new Request("https://example.com/auth/passkey/sign-in/finish",{method:"POST",headers:{"x-forwarded-for":"untrusted"}});
    const issuer=vi.fn<Parameters<PasskeySignIn>,ReturnType<PasskeySignIn>>(async(input)=>{
     expect(input).toMatchObject({userId:"user",signCount:1,lifetimeSeconds:3600,userAgent:"agent",ip:"untrusted"});
     expect([...input.credentialId]).toEqual([...credentialId]);
     expect([...input.publicKey]).toEqual([...publicKey]);
     expect(input.now()).toBe(h.clock.now);
     expect(input.request).toBe(mode==="direct" ? undefined : request);
     if(mode==="denied") throw new AuthError("invalid_credential","Sign-in not allowed");
     if(mode==="unavailable") throw new AuthError("session_unavailable","Session temporarily unavailable");
     if(mode==="failure") throw new Error("private host failure");
     if(mode==="empty") return {sessionToken:"",user:{id:"user",email:""}};
     return {sessionToken:"host-token",user:{id:mode==="wrong-user"?"different":"user",email:""}};
    });
    const auth=createAuth({rpId:"example.com",origins:["https://example.com"],session:{lifetimeSeconds:3600},email:{sendOtp:h.sendOtp},users:{findOrCreateByEmail:h.findOrCreateByEmail},...(mode==="default"?{}:{passkey:{signIn:issuer}})},{db:h.db,deps:h.deps});
    const start=await auth.beginPasskeySignIn();
    const input={...start,credential,userAgent:"agent",ip:"untrusted",...(mode==="direct"?{}:{request})};
    const result=auth.finishPasskeySignIn(input);
    if(["success","direct","default"].includes(mode)) {
      const value=await result;expect(value.user.id).toBe("user");
      if(mode!=="default")expect(value.sessionToken).toBe("host-token");
    } else if(mode==="denied"||mode==="invalid") await expect(result).rejects.toMatchObject({code:"invalid_credential"});
    else if(mode==="unavailable") await expect(result).rejects.toMatchObject({code:"session_unavailable"});
    else if(mode==="failure") await expect(result).rejects.toThrow("private host failure");
    else await expect(result).rejects.toThrow("passkey issuer returned invalid result");
    expect(issuer).toHaveBeenCalledTimes(mode==="default"||mode==="invalid"?0:1);
    expect(getPasskeyByCredentialId(h.db,credentialId)?.signCount).toBe(mode==="default"?1:0);
    expect(h.db.prepare("SELECT COUNT(*) AS n FROM auth_sessions").get()).toEqual({n:mode==="default"?1:0});
    await expect(auth.finishPasskeySignIn(input)).rejects.toThrow();
    expect(issuer).toHaveBeenCalledTimes(mode==="default"||mode==="invalid"?0:1);
   } finally {h.db.close();}
  });
 }
});
